### DISCUSSION TITLE
Test cases for custom Larastan extensions

### DISCUSSION THREAD

**Discussion by mad-briller** [larastan/larastan#1422] [Q&A]
When creating custom rules that require Larastan to inspect the container or configuration of the Laravel application being analysed, is it possible to somehow modify the state of the application before analysis in the test is ran?

from my brief understanding, Larastan is run separately when running tests and as such the naive approach of `app('config')->set()` doesn't work, as the newly spawned process will not see those changes.

**Comment by canvural** [comment]
> Larastan is run separately when running tests 

This is only true for our own tests. (which is deprecated. New rule tests are using `PHPStan\Testing\RuleTestCase`) 

If you are writing a new package that provides rules, your test class should extend `PHPStan\Testing\RuleTestCase` and you'd write your tests there. And how you simulate config depends on what kind of rule you are writing. You could maybe make it a config parameter in `phpstan.neon` file. I'd need more information to be able to help.

**Reply by mad-briller** [reply]
that's good to know and makes a lot of sense.

then i believe the issue is that because i'm trying to use `app()` before running `$this->analyse()`, the Laravel application isn't actually booted yet so i get a `BindingResolutionException` for `config`:

```php
<?php

namespace Tests\Unit\Phpstan;

use Hallnet\LaravelWarehousing\Phpstan\Rules\ManagerRule;
use PHPStan\Rules\Rule;
use PHPStan\Testing\RuleTestCase;

/**
 * @covers \Hallnet\LaravelWarehousing\Phpstan\Rules\ManagerRule
 *
 * @extends RuleTestCase<ManagerRule>
 */
class ManagerRuleTest extends RuleTestCase
{
    protected function getRule(): Rule
    {
        return new ManagerRule();
    }

    public function testCallMethod(): void
    {
        $this->analyse([__DIR__ . '/data/missing-model-config.php'], [
            [
                'Model `MissingModelConfig\Model` not found in warehousing.warehouses configuration.',
                29
            ],
        ]);
    }

    public function testCallMethodWithConfig(): void
    {
        app('config')->set('warehousing.warehouses.' . \MissingModelConfig\Model::class, []);

        $this->analyse([__DIR__ . '/data/missing-model-config.php'], []);
    }

    public static function getAdditionalConfigFiles(): array
    {
        return [
            __DIR__ . '/../../../extension.neon',
            __DIR__ . '/../../../vendor/nunomaduro/larastan/extension.neon',
        ];
    }
}
```

in the above test case `testCallMethod` works fine, but the `testCallMethodWithConfig` fails with the `BindingResolutionException`. I'm trying to test that this rule correctly picks up on a misconfigured system

thanks for your time

**Reply by canvural** [reply]
Hmm. Yeah this is a hard case. This is not static analysis anymore 😄 

I don't have any real solution. But one hacky idea could be that you can create your config file that you want to test with, in the `data` folder. Then before calling `$this->analyse` you can copy that file to `vendor/orchestra/testbench-core/laravel/config` This is the place where dummy Laravel app is booted with Testbench. This should do the trick. And after you'd delete that file to not interfere with other tests. Either in `tearDown` or directly after `$this->analyse`

**Reply by mad-briller** [reply]
> Hmm. Yeah this is a hard case. This is not static analysis anymore

While it's not the _ideal_ way of implementing static analysis, the power of proving correctness in this way cannot be denied, and we've decided that this step away from traditional static analysis is valuable for us.
We have many internal packages that extend Laravel powered by configuration files that developers can use at their own will, so guiding developers towards a correct configuration is very useful when combined with editor plugins and automated pipelines, as with `phpstan/extension-installer` we can write an extension once and gain confidence everywhere they are used, even retroactively!

Larastan also walks a fine line between static and non-static analysis itself in many places :sweat_smile:, and i personally believe it does so for the right reasons. Laravel's magic makes it quite hard to statically analyse things anyway as i'm sure you are very familiar with :D

thanks for your input @canvural, it's a decent idea and an avenue i can investigate

**Reply by mad-briller** [reply]
thinking about it, i could avoid the need for the `app('config')` stuff altogether by using a collector, as inspired by your unused-view pull request :thinking:

**Reply by canvural** [reply]
If it suits your needs, yeah! It's an awesome new feature!
